28課題_Thymeleafを使ったPOST処理

javachead opened the pull request:
**変更内容**

①コントローラー
　・生徒情報を登録する処理の追加
　・コース情報を登録する処理の追加
②サービス
　・新しく登録する生徒のidを決定する処理の追加
　・新しく登録するコースのidを決定する処理の追加
③リポジトリ
　・生徒情報のINSERT文の追加
　・コース情報のINSERT文の追加

＜入力画面＞
![スクリーンショット 2025-01-10 154431](https://github.com/user-attachments/assets/5ca3f96f-aec8-40b4-b29d-6aa3abbda438)

＜結果＞
![スクリーンショット 2025-01-10 154348](https://github.com/user-attachments/assets/cf4511c2-5bb7-43be-b0f1-395c425d2694)

＜DB＞
![スクリーンショット 2025-01-10 162110](https://github.com/user-attachments/assets/94ab798d-6f57-41f5-9d56-71e58085e421)
![スクリーンショット 2025-01-10 162304](https://github.com/user-attachments/assets/8d450d4a-6023-4498-bf36-d05e39993b0a)
![スクリーンショット 2025-01-10 183752](https://github.com/user-attachments/assets/d4d88c4a-36d6-45cf-ac92-654008b4f3a1)
![スクリーンショット 2025-01-10 193441](https://github.com/user-attachments/assets/bd6242a9-92be-4d15-8081-f243fd6ff68d)

Commit:
①データベースアクセスの仕組みを JPA から MyBatis に変更

Commit:
②
・学生詳細情報に、受講生IDが紐づけされたコース情報のみを反映するよう変更
・不要なHibernateの設定を削除

Commit:
28課題/学生管理を強化し、コース統合を実現

`StudentCoursesMapper` と `StudentMapper` を導入し、データ処理を効率化。学生とコースの保存メソッドや、Thymeleafテンプレートの更新、Spring Bootによるフォームバリデーションを実装。設定とログ機能も改善。

KeitaUenishi commented on src/main/java/raisetech/student/data/StudentCourses.java:
```diff
@@ -0,0 +1,36 @@
+package raisetech.student.data;
+
+import jakarta.persistence.*;
```
`*` ワイルドカードでのimportはなるべくしない方がいいです
import対象は明確にした方が、何を使用するかが明示的になりソースコードの保守性も向上します

KeitaUenishi commented on src/main/java/raisetech/student/data/StudentCourses.java:
```diff
@@ -0,0 +1,36 @@
+@Table(name = "studentscourses")
+@Data
+@NoArgsConstructor
+public class StudentCourses {
```
`StudentCourses`
単体のオブジェクトになるので、複数形の`s` は不要かと思います

KeitaUenishi replied:
こちらですが、クラスはあえて削除せず残されている形でしょうか？

javachead replied:
消し忘れです...お恥ずかしい限りです

KeitaUenishi replied:
なるほどです！このPRでは残していく方針ですかね？
https://github.com/javachead/ServiceController/pull/4/commits/690495f9041c631dbca743eb5d26c5db5bd296bd
こちらのコミットで削除確認できずのため…

KeitaUenishi commented on src/main/java/raisetech/student/repository/StudentCourseRepository.java:
```diff
@@ -1,16 +1,13 @@
-    List<StudentCourse> findAllStudentCourses();
+@Mapper
+public interface StudentCourseRepository {
+    @Select("SELECT * FROM studentscourses WHERE studentscourses.student_id = #{studentId}")
```
FROM句のテーブル名がDBと異なっている気がしますが、こちらでデータ取得できますか？

javachead replied:
あとからDBの名前を正しいstudents_courses に変えました、お伝えするコメント入れわすれてしまい申し訳ございません。

KeitaUenishi commented on src/main/java/raisetech/student/repository/StudentMapper.java:
こちらも、RepositoryとMapperとしてInterfaceの定義を分けられているのは理由がありますか？

javachead replied:
保守性が上がるという考えでわけました。

KeitaUenishi commented on src/main/resources/templates/registrationSuccess.html:
成功画面いれてるのいいですね！ 👍

KeitaUenishi commented on src/main/java/raisetech/student/controller/StudentController.java:
```diff
@@ -1,59 +1,98 @@
+            // メソッド名を修正: saveStudentAndCourses を呼び出す
+            studentService.saveStudentAndCourses(student, courses);
+
+            logger.info("学生情報とコース情報を登録しました: {}", student.getName());
```
しっかりロギングされている…素晴らしい……

javachead replied:
ロギング今後もしっかり入れられるように頑張りたいです！

javachead commented on src/main/java/raisetech/student/repository/StudentCoursesMapper.java:
RepositoryとMapperを分けた理由としては、責務を明確に分離し、保守性を高めることを意図していました。
今後は基本的にRepositoryにMapperを定義するようにします。
ありがとうございます。

javachead replied:
かしこまりました。ありがとうございます。

javachead commented on src/main/java/raisetech/student/data/StudentCourses.java:
```diff
@@ -0,0 +1,36 @@
+    private Long id;
+
+    // 明示的にセッターメソッドを追加
+    @Setter
```
@Dateではなく@Setterに書き換えわすれてしまいました💦
何を使用するかが明示的になり、ソースコードの保守性も向上するとの目的で@Date使用はできるだけ避けるべきとありましたので、@Getterや@Setterを出来るだけ使う方向で進めています。
@Dateに関する考え方もワイルドカード同様の考え方でよろしいでしょうか。

KeitaUenishi replied:
なるほど！
> @Dateに関する考え方もワイルドカード同様の考え方でよろしいでしょうか。

おっしゃる通りですね！
使う使わないは意見分かれると思いますが、あまりブラックボックスになりすぎないように明示的にしておく方が分かり良いですねー

javachead commented on src/main/java/raisetech/student/data/StudentCourses.java:
```diff
@@ -0,0 +1,36 @@
+
+    // 明示的にセッターメソッドを追加
+    @Setter
+    @ManyToOne
+    @JoinColumn(name = "student_id", nullable = false)
```
ありがとうございます。

Commit:
28課題/修正2
①StudentCourseへ命名変更

Commit:
28課題/修正
①RepositoryとMapperの統合→Repositorへ
②@Dataを違う方法で実装
③JPA使用時のコードを削除
④* ワイルドカードでのimport分の修正
⑤StudentCourseへ命名変更
⑥DBの名前を正しいstudent_coursesへ変更

KeitaUenishi commented on logs/mybatis.log/spring.log:
logのディレクトリですが、.gitignoreに入れてバージョン管理からは外しておいた方がいいですねー
